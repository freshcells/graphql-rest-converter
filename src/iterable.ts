export class AsyncQueue<T> {
  private queue: (T | null | Error)[] = []
  processedItems: T[] = []
  #resolve: false | ((item: T | null | Error) => void) = false
  #reject: false | ((msg?: Error) => void) = false

  /**
   * alias for `add(null)`
   */
  terminate() {
    this.add(null)
  }

  /**
   * alias for `add(error)`
   */
  reject(error: Error) {
    this.add(error)
  }

  add(item: T | null | Error) {
    if (this.#resolve && !(item instanceof Error)) {
      this.#resolve(item)
      if (item !== null) {
        this.processedItems.push(item)
      }
      this.#resolve = false
      return
    }
    if (this.#reject && item instanceof Error) {
      this.#reject(item)
      this.#reject = false
      return
    }
    this.queue.push(item)
  }

  async *[Symbol.asyncIterator]() {
    while (true) {
      const result = await new Promise((resolve, reject) => {
        if (this.queue.length > 0) {
          const nextItem = this.queue.shift()
          if (nextItem instanceof Error) {
            return reject(nextItem)
          }
          if (nextItem) {
            this.processedItems.push(nextItem)
          }
          return resolve(nextItem)
        }
        this.#resolve = resolve
        this.#reject = reject
      })
      if (result === null) {
        return
      }
      yield result as T
    }
  }
}
