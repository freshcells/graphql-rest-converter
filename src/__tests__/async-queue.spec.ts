import { AsyncQueue } from '../iterable.js'

describe('AsyncQueue', () => {
  it('should allow to iterate over values', async () => {
    const queue = new AsyncQueue<number>()
    const p1 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.add(1)
        resolve()
      }, 5),
    )
    const p2 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.add(2)
        resolve()
      }, 15),
    )
    const p3 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.add(3)
        resolve()
      }, 20),
    )
    const p4 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.terminate()
        resolve()
      }, 30),
    )
    const allItems = []
    for await (const item of queue) {
      allItems.push(item)
    }
    await Promise.all([p1, p2, p3, p4])
    expect(queue.processedItems).toEqual(allItems)
    expect(allItems).toMatchSnapshot()
  })

  it('should throw in case of errors', async () => {
    const queue = new AsyncQueue<number>()
    const p1 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.add(1)
        resolve()
      }, 5),
    )
    const p2 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.reject(new Error('invalid process'))
        resolve()
      }, 15),
    )
    await expect(async () => {
      const allItems = []
      for await (const item of queue) {
        allItems.push(item)
      }
      await Promise.all([p1, p2])
    }).rejects.toThrow(new Error('invalid process'))
  })

  it('should work for async work', async () => {
    const queue = new AsyncQueue<{ fnc: () => Promise<number> }>()
    const p1 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.add({
          fnc: async () => {
            return new Promise((resolve) => {
              setTimeout(() => {
                resolve(1)
              }, 10)
            })
          },
        })
        resolve()
      }, 5),
    )

    const p2 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.add({
          fnc: async () => {
            return new Promise((resolve) => {
              setTimeout(() => {
                resolve(2)
              }, 10)
            })
          },
        })
        resolve()
      }, 10),
    )

    const p3 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.terminate()
        resolve()
      }, 10),
    )

    const results = []

    for await (const item of queue) {
      results.push(await item.fnc())
    }

    await Promise.all([p1, p2, p3])
    expect(results).toMatchSnapshot()
    expect(queue.processedItems).toHaveLength(results.length)
  })
  it('should work for errors in async work', async () => {
    const queue = new AsyncQueue<{ fnc: () => Promise<number> }>()
    const p1 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.add({
          fnc: async () => {
            return new Promise((resolve) => {
              setTimeout(() => {
                resolve(1)
              }, 10)
            })
          },
        })
        resolve()
      }, 5),
    )

    const p2 = new Promise<void>((resolve) =>
      setTimeout(() => {
        queue.add({
          fnc: async () => {
            return new Promise((resolve, reject) => {
              setTimeout(() => {
                reject(new Error('Something is broken'))
              }, 10)
            })
          },
        })
        resolve()
      }, 10),
    )

    await expect(async () => {
      const results = []
      for await (const item of queue) {
        results.push(await item.fnc())
      }
      await Promise.all([p1, p2])
    }).rejects.toThrow(new Error('Something is broken'))
  })
  it('should allow direct termination', async () => {
    const noopQueue = new AsyncQueue()
    noopQueue.terminate()
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    for await (const item of noopQueue) {
      // ..
    }
    expect(true)
  })
})
