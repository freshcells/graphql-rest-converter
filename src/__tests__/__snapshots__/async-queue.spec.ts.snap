// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncQueue should allow to iterate over values 1`] = `
[
  1,
  2,
  3,
]
`;

exports[`AsyncQueue should work for async work 1`] = `
[
  1,
  2,
]
`;
